linter: junit formatter contains last file diagnistics only

### What version of Oxlint are you using?

1.31.0

### What command did you run?

`oxlint -c .oxintrc.json --format junit`

### What does your `.oxlintrc.json` config file look like?
N/A

### What happened?

consider this default oxlint output for a given two file project
```
> pnpm oxlint *.tsx -D all

  × eslint(curly): Expected { after 'if' condition.
   ╭─[bar.tsx:4:22]
 3 │ const DummyComponent: React.FC = () => {
 4 │     if (ref.current) ref.current.focus()
   ·                      ───────────────────
 5 │ 
   ╰────
  help: Replace `ref.current.focus()` with `{ref.current.focus()}`.

  × eslint(curly): Expected { after 'if' condition.
   ╭─[foo.tsx:7:22]
 6 │   useEffect(() => {
 7 │     if (ref.current) ref.current.focus()
   ·                      ───────────────────
 8 │   }, [])
   ╰────
  help: Replace `ref.current.focus()` with `{ref.current.focus()}`.

  × eslint(no-constant-binary-expression): Unexpected constant nullishness on the left-hand side of a "??" expression
    ╭─[foo.tsx:11:11]
 10 │       return (
 11 │ ╭─▶     <div>{Boolean(ref.current) ?? (
 12 │ │         <input type="text" ref={ref} />
 13 │ ╰─▶     )}
 14 │         </div>
    ╰────
  help: This expression always evaluates to the constant on the left-hand side

Found 0 warnings and 3 errors.
Finished in 40ms on 2 files with 335 rules using 12 threads.
```

It has simillar correct outputs with other formatters
<details><summary>stylish</summary>

```
> pnpm oxlint *.tsx -D all --format stylish

/tmp/tmp.lLodRLQny0/foo.tsx
  7:22   error  Expected { after 'if' condition.  eslint(curly)
  11:11  error  Unexpected constant nullishness on the left-hand side of a "??" expression  eslint(no-constant-binary-expression)

/tmp/tmp.lLodRLQny0/bar.tsx
  4:20  error  Expected { after 'if' condition.  eslint(curly)
```

</details>

<details>
<summary>Json</summary>

```
> pnpm oxlint *.tsx -D all --format json
{ "diagnostics": [{"message": "Expected { after 'if' condition.","code": "eslint(curly)","severity": "error","causes": [],"url": "https://oxc.rs/docs/guide/usage/linter/rules/eslint/curly.html","help": "Replace `ref.current.focus()` with `{ref.current.focus()}`.","filename": "foo.tsx","labels": [{"span": {"offset": 178,"length": 19,"line": 7,"column": 22}}],"related": []},
{"message": "Unexpected constant nullishness on the left-hand side of a \"??\" expression","code": "eslint(no-constant-binary-expression)","severity": "error","causes": [],"url": "https://oxc.rs/docs/guide/usage/linter/rules/eslint/no-constant-binary-expression.html","help": "This expression always evaluates to the constant on the left-hand side","filename": "foo.tsx","labels": [{"span": {"offset": 229,"length": 69,"line": 11,"column": 11}}],"related": []},
{"message": "Expected { after 'if' condition.","code": "eslint(curly)","severity": "error","causes": [],"url": "https://oxc.rs/docs/guide/usage/linter/rules/eslint/curly.html","help": "Replace `ref.current.focus()` with `{ref.current.focus()}`.","filename": "bar.tsx","labels": [{"span": {"offset": 94,"length": 19,"line": 4,"column": 20}}],"related": []}],
              "number_of_files": 2,
              "number_of_rules": 335,
              "threads_count": 12,
              "start_time": 0.029430435
            }
              
```

</details>

<details>
<summary>Checkstyle</summary>

```
> pnpm oxlint *.tsx -D all --format checkstyle
<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name="foo.tsx"><error line="7" column="22" severity="error" message="Expected { after &apos;if&apos; condition." source="eslint(curly)" /><error line="11" column="11" severity="error" message="Unexpected constant nullishness on the left-hand side of a &quot;??&quot; expression" source="eslint(no-constant-binary-expression)" /></file> <file name="bar.tsx"><error line="4" column="20" severity="error" message="Expected { after &apos;if&apos; condition." source="eslint(curly)" /></file></checkstyle>
```

</details> 




However for the junit output - it only has a **singular** error, and for only a singular file

```
> pnpm oxlint *.tsx -D all --format junit
<?xml version="1.0" encoding="UTF-8"?>
<testsuites name="Oxlint" tests="3" failures="0" errors="3">
    <testsuite name="bar.tsx" tests="1" disabled="0" errors="1" failures="0">
        <testcase name="eslint(curly)">
            <error message="Expected { after &apos;if&apos; condition.">line 4, column 20, Expected { after &apos;if&apos; condition.</error>
        </testcase>
    </testsuite>
</testsuites>
```

Where we would instead expect multiple testSuites and multiple `errors`

I have a PR that might fix this which will raise shortly - but needs someone to check it :D

--- a/apps/oxlint/src/output_formatter/junit.rs
+++ b/apps/oxlint/src/output_formatter/junit.rs
@@ -75,9 +75,8 @@ fn format_junit(diagnostics: &[Error]) -> String {
                 description,
                 severity
             );
-            let test_case = format!(
-                "\n        <testcase name=\"{rule}\">\n{status}\n        </testcase>"
-            );
+            let test_case =
+                format!("\n        <testcase name=\"{rule}\">\n{status}\n        </testcase>");
             test_cases.push_str(&test_case);
         }
         test_suites.push(format!(

--- a/apps/oxlint/src/output_formatter/junit.rs
+++ b/apps/oxlint/src/output_formatter/junit.rs
@@ -33,18 +33,21 @@ impl DiagnosticReporter for JUnitReporter {
 
 fn format_junit(diagnostics: &[Error]) -> String {
     let mut grouped: FxHashMap<String, Vec<&Error>> = FxHashMap::default();
-    let mut total_errors = 0;
-    let mut total_warnings = 0;
 
     for diagnostic in diagnostics {
         let info = Info::new(diagnostic);
         grouped.entry(info.filename).or_default().push(diagnostic);
     }
 
-    let mut test_suite = String::new();
-    for diagnostics in grouped.values() {
-        let diagnostic = diagnostics[0];
-        let filename = Info::new(diagnostic).filename;
+    let mut filenames: Vec<_> = grouped.keys().cloned().collect();
+    filenames.sort();
+
+    let mut total_errors = 0;
+    let mut total_warnings = 0;
+    let mut test_suites = Vec::new();
+
+    for filename in filenames {
+        let diagnostics = grouped.get(&filename).expect("filename collected from map");
         let mut test_cases = String::new();
         let mut error = 0;
         let mut warning = 0;
@@ -72,25 +75,26 @@ fn format_junit(diagnostics: &[Error]) -> String {
                 description,
                 severity
             );
-            let test_case =
-                format!("\n        <testcase name=\"{rule}\">\n{status}\n        </testcase>");
-            test_cases = format!("{test_cases}{test_case}");
+            let test_case = format!(
+                "\n        <testcase name=\"{rule}\">\n{status}\n        </testcase>"
+            );
+            test_cases.push_str(&test_case);
         }
-        test_suite = format!(
+        test_suites.push(format!(
             "    <testsuite name=\"{}\" tests=\"{}\" disabled=\"0\" errors=\"{}\" failures=\"{}\">{}\n    </testsuite>",
             filename,
             diagnostics.len(),
             error,
             warning,
             test_cases
-        );
+        ));
     }
     let test_suites = format!(
         "<testsuites name=\"Oxlint\" tests=\"{}\" failures=\"{}\" errors=\"{}\">\n{}\n</testsuites>\n",
         total_errors + total_warnings,
         total_warnings,
         total_errors,
-        test_suite
+        test_suites.join("\n")
     );
 
     format!("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n{test_suites}")
@@ -125,6 +129,40 @@ mod test {
         let warning = OxcDiagnostic::warn("warning message")
             .with_label(Span::new(0, 9))
             .with_source_code(NamedSource::new("file.js", "debugger;"));
+
+        reporter.render_error(error);
+        reporter.render_error(warning);
+
+        let output = reporter.finish(&DiagnosticResult::default()).unwrap();
+        assert_eq!(output, EXPECTED_REPORT);
+    }
+
+    #[test]
+    fn test_junit_reporter_multiple_files() {
+        const EXPECTED_REPORT: &str = r#"<?xml version="1.0" encoding="UTF-8"?>
+<testsuites name="Oxlint" tests="2" failures="1" errors="1">
+    <testsuite name="a.js" tests="1" disabled="0" errors="1" failures="0">
+        <testcase name="">
+            <error message="error message a">line 1, column 1, error message a</error>
+        </testcase>
+    </testsuite>
+    <testsuite name="b.js" tests="1" disabled="0" errors="0" failures="1">
+        <testcase name="">
+            <failure message="warning message b">line 1, column 1, warning message b</failure>
+        </testcase>
+    </testsuite>
+</testsuites>
+"#;
+
+        let mut reporter = JUnitReporter::default();
+
+        let error = OxcDiagnostic::error("error message a")
+            .with_label(Span::new(0, 8))
+            .with_source_code(NamedSource::new("a.js", "let a = ;"));
+
+        let warning = OxcDiagnostic::warn("warning message b")
+            .with_label(Span::new(0, 9))
+            .with_source_code(NamedSource::new("b.js", "debugger;"));
 
         reporter.render_error(error);
         reporter.render_error(warning);

